crates/uv-pep508/src/lib.rs

```diff
@@ -251,6 +252,52 @@ impl<T: Pep508Url> Serialize for Requirement<T> {
     }
 }
 
+impl<T: Pep508Url> CacheKey for Requirement<T>
+where
+    T: Display,
+{
+    fn cache_key(&self, state: &mut CacheKeyHasher) {
+        self.name.as_str().cache_key(state);
+
+        self.extras.len().cache_key(state);
+        for extra in &self.extras {
+            extra.as_str().cache_key(state);
+        }
+
+        // TODO(zanieb): We inline cache key handling for the child types here, but we could
+        // move the implementations to the children. The intent here was to limit the scope of
+        // types exposing the `CacheKey` trait for now.
+        if let Some(version_or_url) = &self.version_or_url {
+            1u8.cache_key(state);
+            match version_or_url {
+                VersionOrUrl::VersionSpecifier(spec) => {
+                    0u8.cache_key(state);
+                    spec.len().cache_key(state);
+                    for specifier in spec.iter() {
+                        specifier.operator().as_str().cache_key(state);
+                        specifier.version().to_string().cache_key(state);
```

Comment:
It could make sense to impl it for `Version` if you feel motivated, to avoid this allocation.